apps/emqx/test/emqx_routing_SUITE.erl

```diff
@@ -307,8 +303,92 @@ t_slow_rlog_routing_consistency(Config) ->
         ct:fail("quick re-subscribe failed to add a route")
     end.
 
+t_cleanup_routes_race(init, Config) ->
+    ok = snabbkaffe:start_trace(),
+    [Core1, Core2, _Repl] = ?config(cluster, Config),
+    lists:foreach(
+        fun(N) -> mock_router_helper(N, 4000) end,
+        [Core1, Core2]
+    ),
+    Config;
+t_cleanup_routes_race('end', Config) ->
+    ok = snabbkaffe:stop(),
+    [Core1, Core2, _Repl] = ?config(cluster, Config),
+    lists:foreach(
+        fun unmock_router_helper/1,
+        [Core1, Core2]
+    ).
+
+t_cleanup_routes_race(Config) ->
+    ?check_trace(
+        begin
+            [_Core1, _Core2, Replicant] = Nodes = ?config(cluster, Config),
+            [_, _, ReplSpec] = ?config(node_specs, Config),
+
+            Topic = <<"t/test/must_be_cleaned">>,
+            Topic1 = <<"t/testmust_NOT_be_cleaned">>,
+            ok = rpc:call(Replicant, emqx_router, do_add_route, [Topic]),
+            wait_for_route(Replicant, Topic, Replicant),
+            {ok, SubRef} = snabbkaffe:subscribe(
+                ?match_event(#{?snk_kind := emqx_router_helper_cleanup_done}),
+                %% Currently, router_helper receives two nodedown messages per a down replicant,
+                %% One is by its own node monitor and one from Mria.
+                %% We want to wait for all these cleanups to finish on every out of 2 core nodes.
+                _NEvents = 4,
+                _Timeout = infinity
+            ),
+
+            emqx_cth_cluster:restart(Replicant, ReplSpec),
+            ok = rpc:call(Replicant, emqx_router, do_add_route, [Topic1]),
+
+            %% wait for Mria route replication,
+            %% cannot poll has_route, because late cleanup may wipe it away
+            %% if the test is about to fail due to a race
+            timer:sleep(1000),
```

Comment:
Q: Why is this sleep needed?

Reply:
To wait for core -> replicant  new route replication, the comment above explains it (though probably it is not very clear..). 
The test case assumes that cleanups may ocur after node is up again, so the cleanups can remove new routes. 
The expected order of events:
1. restart replicant
2. add new route on the replicant
3. wait a little to give some time to a core node to replicat this new route back to the replicant
4. make sure all cleanups have been done (via snabkaffe)
5. check that every node has the new route but doesn't have the old one.